lua/entities/gmod_tardis/modules/sh_artronenergy.lua

```diff
@@ -0,0 +1,311 @@
+    end)
+
+    ENT:AddHook("TardisControlUsed", "artron", function(self, control)
+        if not TARDIS:GetSetting("artron_energy") then return end
+
+        if artron_values.cost_controls[control] then
+            self:AddArtron(artron_values.cost_controls[control])
+        end
+    end)
+
+    ENT:AddHook("HandleNoDemat", "artron", function(self)
+        if not TARDIS:GetSetting("artron_energy") then return end
+
+        if ArtronDematCheck(self) == true then return end
+        self:AddArtron(artron_values.cost_failed_demat)
+    end)
+
+    ENT:AddHook("DematStart", "artron", function(self)
+        if not TARDIS:GetSetting("artron_energy") then return end
+        if self:CallHook("ShouldUpdateArtron") == false then return end
+
+        if self:GetData("demat-fast", false) then
+            self:AddArtron(artron_values.cost_full)
+        else
+            self:AddArtron(artron_values.cost_demat)
+        end
+    end)
+
+    ENT:AddHook("MatStart", "artron", function(self)
+        if not TARDIS:GetSetting("artron_energy") then return end
+        if self:CallHook("ShouldUpdateArtron") == false then return end
+
+        if self:GetData("demat-fast",false) ~= true then
+            self:AddArtron(artron_values.cost_mat)
+        end
+    end)
+
+    ENT:AddHook("HADSTrigger", "artron", function(self)
+        if not TARDIS:GetSetting("artron_energy") then return end
+
+        if self:GetData("artron-val",0) < -artron_values.cost_hads then
+            self:SetArtron(1)
+            return
+        end
+        self:AddArtron(artron_values.cost_hads)
+    end)
+
+
+
+
+    --
+    -- Conditions for artron changing
+    --
+
+    ENT:AddHook("CanIncreaseArtron", "interrupt-cooldown", function(self)
+        if self:GetData("teleport-interrupted") then
+            return false
+        end
+    end)
+
+    ENT:AddHook("ShouldUpdateArtron", "repair", function(self)
+        if self:GetData("repair-primed") or self:GetData("repairing") then
+            return false
+        end
+    end)
+
+    ENT:AddHook("ShouldUpdateArtron", "redecoration", function(self)
+        if self:GetData("redecorate")
+            or self:GetData("redecorate_parent")
+            or self:GetData("redecorate_child")
+            or self:GetData("is_redecorate_child")
+        then
+            return false
+        end
+    end)
+
+    ENT:AddHook("ShouldUpdateArtron", "hads", function(self)
+        if self:GetData("hads-triggered")
+            or self:GetData("hads-need-remat")
+            or self:GetData("hads-auto-remat")
+        then
+            return false
+        end
+    end)
+
+    ENT:AddHook("SettingChanged", "maxartron-changed", function(self, id, val)
+        if id ~= "artron_energy_max" then return end
+
+        if self:GetData("artron-val",0) > val then
+            self:SetArtron(val)
+        end
+    end)
+
+
+    --
+    -- Emergency artron behaviour
+    --
+    function ENT:ForceAddArtron()
```

Comment:
This works even when health is disabled, what should we do in that scenario? The notification implies health has been decreased to compensate even if it hasn't. Additionally it is possible to put the TARDIS into a health warning mode even with health disabled using this control,

Calls to SetHealth should do nothing when it's turned off and turning it off should max health out like we do for adjusting the artron max energy

Reply:
b390b71908ed317de4c4c2685f64bc68ffec0cf1
I've done some safeguards for that both for Engine Release and for changing health in general

Reply:
A more general point I suppose is should enabling artron energy even be possible if health is disabled, is that a scenario we want to support?

Reply:
Why not?
In all scenarios except for engine release, artron energy is independent from health
It's not like we can't use one without the other
(you just have to be more attentive about when you want to materialise in the vortex, but it still works)

Reply:
That's fine, we just need to make sure as part of testing that it works correctly when health is disabled